tests/test_core.py

```diff
@@ -103,3 +111,114 @@ def test_distance_matrix(p: int, dtype: torch.dtype) -> None:
         rtol=rtol,
         atol=atol,
     )
+
+
+def test_quantisation_cpu() -> None:
+    assert set(
+        pea.quantise_fpx(
+            torch.linspace(-4, 4, steps=100),
+            exponent_bits=2,
+            mantissa_bits=1,
+            rounding="nearest",
+        ).tolist()
+    ) == {sx for x in [0, 0.25, 0.5, 0.75, 1, 1.5, 2, 3] for sx in [x, -x]}
+
+    assert set(
+        pea.quantise_fpx(
+            torch.linspace(-10, 10, steps=1000),
+            exponent_bits=3,
+            mantissa_bits=0,
+            rounding="nearest",
+        )
+        .abs()
+        .tolist()
+    ) == {0, 0.125, 0.25, 0.5, 1, 2, 4, 8}
+
+
+@pytest.mark.parametrize("device", ["cpu", "ipu"])
+def test_quantisation_stochastic_rounding(device: str) -> None:
+    if device == "ipu" and not poptorch.ipuHardwareIsAvailable():
+        pytest.skip("quantise_fpx() requires IPU hardware")
+
+    n = 10000
+    x, grad_y = -1.35, 3.0
+    out = run_forward_and_backward(
+        lambda x: dict(
+            y=pea.quantise_fpx(
+                x, exponent_bits=2, mantissa_bits=1, rounding="stochastic", bwd=True
+            )
+        ),
+        dict(x=torch.full((n,), x)),
+        grad_outputs=dict(y=torch.full((n,), grad_y)),
+        device=device,
+    )
+
+    y_count = collections.Counter(out["y"].tolist())
+    assert y_count.keys() == {-1.5, -1.0}
+    expected_ratio = (1.35 - 1.0) / 0.5
+    nearest_ratio = y_count[-1.5] / sum(y_count.values())
+    std_x3 = 3 * (expected_ratio * (1 - expected_ratio) / n) ** 0.5
+    assert expected_ratio - std_x3 < nearest_ratio < expected_ratio + std_x3
+
+    assert collections.Counter(out["grad_x"].tolist()) == {3.0: n}
+
+
+@pytest.mark.parametrize("device", ["cpu", "ipu"])
+def test_quantisation_variants(device: str) -> None:
+    if device == "ipu" and not poptorch.ipuHardwareIsAvailable():
+        pytest.skip("quantise_fpx() requires IPU hardware")
+
+    # Note: some gymnastics here to get everything into one test, for efficiency
+    def _fn(**args: Tensor) -> Dict[str, Tensor]:
+        return {
+            f"y{suffix}": quantise(  # type:ignore[operator]
+                args[f"x{suffix}"],
+                exponent_bits=5,
+                mantissa_bits=2,
+                rounding="nearest",
+            )
+            for suffix, quantise in [
+                ("", pea.quantise_fpx),
+                ("_ste", pea.quantise_fpx_ste),
+                ("_grad_only", pea.quantise_fpx_grad),
+            ]
+        }
+
+    x = torch.linspace(-1e5, 1e5, steps=1000)
+    grad_y = torch.flip(x, (0,))
+    out = run_forward_and_backward(
+        _fn,
+        {name: x for name in ["x", "x_ste", "x_grad_only"]},
+        grad_outputs={name: grad_y for name in ["y", "y_ste", "y_grad_only"]},
+        device=device,
+    )
+
+    def assert_quantised(t: Tensor) -> None:
+        assert len(set(t.tolist())) <= 256
+        torch.testing.assert_close(t.max(), torch.tensor(57344.0))
```

Comment:
I think the rtol here for fp32 will be `1.3e-6`, which doesn't seem enough if you assume you might occasionally get less than the max value (maybe you won't though...)

A quick test:
```
>>> x = torch.tensor(57344.0)
>>> y = torch.tensor(49152.)
>>> assert_close(x, y)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/charlieb/Projects/graphcore/unit-scaling/.venv/lib/python3.10/site-packages/torch/testing/_comparison.py", line 1511, in assert_close
    raise error_metas[0].to_error(msg)
AssertionError: Scalars are not close!

Absolute difference: 8192.0 (up to 1e-05 allowed)
Relative difference: 0.16666666666666666 (up to 1.3e-06 allowed)
```
I'd be tempted to set an atol of maybe 2**14 ish?

Reply:
I think I should always get the max, since `x = torch.linspace(-1e5, 1e5, steps=1000)` in this test case. I'm using nearest rounding here, but I think the same should be true if it were stochastic(?)